### DISCUSSION TITLE
How do I load data on the server that depends on user preferences?

### DISCUSSION THREAD

**Discussion by rfgamaral** [remix-run/remix#4150] [Q&A] [answered]
Hi there,

I'm working on something that calls multiple APIs to aggregate some data, and then show that data to the user. However, the APIs that I'm calling depend on a user preference, and I'm not sure where to store that in a way that I can access that preference on the server. Of course, the initial request will probably have a default set of APIs to call, but once the user tweaks the user preferences, I need to store that choice somewhere, and use it on the server-side.

Another important detail is that I don't have user authentication, nor do I have a need for it, so I can't store user preferences in some database, and then read it from the server. In a client-side application, this would normally be done with `localStorage`, but the server does not have access to that.

I understand that cookies are generally the recommended alternative for Remix, but while in this question I'm talking about a single use preference, I might have multiple user preferences in the future that will affect what the server does to retrieve and/or manipulate data, and I'm not sure cookies will scale well. Not to mention that all cookies are sent with every request, increasing the payload on every request as well.

What are my alternatives?

**Comment by sergiodxa** [comment, marked as answer]
With the constraint that you don't have a database, the next best place to store that is using [sessions](https://remix.run/docs/en/v1/api/remix#sessions), so you can create a [cookie session storage](https://remix.run/docs/en/v1/api/remix#createcookiesessionstorage) and store all the user preferences in a cookie, this way your loader and actions will have access to that data.

Using sessions you can store more than one value (using plain cookies too btw), and unless your user preferences are a large JSON file I wouldn't care too much about sending them on every request, user preferences are usually a boolean or string value.

But if, for any reason, you still want to avoid the extra payload on every request you can play with other session storages aside of cookies, like using [file system](https://remix.run/docs/en/v1/api/remix#createfilesessionstorage-node) (if you don't deploy to serverless), use a [memory storage](https://remix.run/docs/en/v1/api/remix#creatememorysessionstorage) (again not serverless and also a single instance), use [Cloudflare KV](https://remix.run/docs/en/v1/api/remix#createcloudflarekvsessionstorage-cloudflare-workers) (if you deploy to Cloudflare) or [Art Table](https://remix.run/docs/en/v1/api/remix#createarctablesessionstorage-architect-amazon-dynamodb) (if you deploy using architect or AWS DynamoDB), finally a last option is to create your own [session storage](https://remix.run/docs/en/v1/api/remix#createsessionstorage) and use any other place to store the session data like Redis or a third party service (e.g. Airtable, Firebase, Supabase, etc.)

**Reply by rfgamaral** [reply]
I guess my only choice is a cookie session storage, because I do intend to deploy to serverless.

Thank you :)

**Reply by rfgamaral** [reply]
@sergiodxa What about using hidden inputs in the form to only send the user settings I need for that particular form, (instead of all user settings that might exist there)? I could read them off `localStorage`, and place them in hidden input during render time.

I can see 2 caveats that I'm not sure if they matter and/or if Remix does something about it:

1. These values would be shown in the query string, since the form method is `GET`;
2. It wouldn't be possible to do SSR with these values in place, and I'm wondering what would happen to the rendering if I went this route. Would Remix still render _something_ on the server side, and the client rendering would fill in the gaps, or everything would be rendered client-side?

I'll probably go with cookie session storage, but I'm just wondering about your thoughts on that alternative.

**Reply by sergiodxa** [reply]
If you don't need that on the initial document request that could be an option.

About your first question, if the preferences are sensible data you would have to use a POST instead of a GET, but if it's sensible data you shouldn't store it on localStorage anyway since that's easily accessible by any JS code.

About your second question, you can have a hook to read from localStorage inside an effect and return `null` on SSR, then use that to render or not the hidden input, so your UI will render as usual on SSS and once the app is hydrated it will read from localStorage and give render the hidden inputs. The main issue here is if the user submits the form before JS load.

**Reply by rfgamaral** [reply]
> If you don't need that on the initial document request that could be an option.

I do, so it's not an option. Thanks for clarifying this for me.
